cpp/include/cudf/contiguous_split.hpp

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace cudf {
+
+/**
+ * @addtogroup column_copy
+ * @{
+ * @file
+ * @brief Table APIs for contiguous_split, pack, unpack, and metadadata
+ */
+
+/**
+ * @brief Column data in a serialized format
+ *
+ * @ingroup copy_split
+ *
+ * Contains data from an array of columns in two contiguous buffers: one on host, which contains
+ * table metadata and one on device which contains the table data.
+ */
+struct packed_columns {
+  /**
+   * @brief Host-side metadata buffer used for reconstructing columns via unpack.
+   *
+   * @ingroup copy_split
+   */
+  struct metadata {
+    metadata() = default;
+
+    /**
+     * @brief Construct a new metadata object
+     *
+     * @param v Host-side buffer containing metadata
+     */
+    metadata(std::vector<uint8_t>&& v) : data_(std::move(v)) {}
+
+    /**
+     * @brief Returns pointer to the host-side metadata buffer data
+     *
+     * @return Pointer to the host-side metadata buffer
+     */
+    [[nodiscard]] uint8_t const* data() const { return data_.data(); }
+
+    /**
+     * @brief Returns size of the metadata buffer
+     *
+     * @return Size of the metadata buffer
+     */
+    [[nodiscard]] size_t size() const { return data_.size(); }
+
+   private:
+    std::vector<uint8_t> data_;
+  };
+
+  packed_columns()
+    : metadata_(std::make_unique<metadata>()), gpu_data(std::make_unique<rmm::device_buffer>())
+  {
+  }
+
+  /**
+   * @brief Construct a new packed columns object
+   *
+   * @param md Host-side metadata buffer
+   * @param gd Device-side data buffer
+   */
+  packed_columns(std::unique_ptr<metadata>&& md, std::unique_ptr<rmm::device_buffer>&& gd)
+    : metadata_(std::move(md)), gpu_data(std::move(gd))
+  {
+  }
+
+  std::unique_ptr<metadata> metadata_;           ///< Host-side metadata buffer
+  std::unique_ptr<rmm::device_buffer> gpu_data;  ///< Device-side data buffer
```

Comment:
We also don't need to wrap the device buffer in a smart pointer.
```suggestion
  rmm::device_buffer gpu_data;  ///< Device-side data buffer
```

Reply:
Such smart pointer is needed as the output of `continuous_split` so it can be used in Python (Cython requires the returned object to be on heap memory). So for simplicity, just take over the content of the input pointers:
```
 packed_columns(std::unique_ptr<metadata>&& md, std::unique_ptr<rmm::device_buffer>&& gd)
    : metadata_(std::move(*md)), gpu_data(std::move(*gd))
  {
  }
```

Reply:
That's OK, no problem.